symengine/basic_to_upoly.cpp

```diff
@@ -0,0 +1,363 @@
+                        RCP<const Pow> powx2
+                            = rcp_static_cast<const Pow>(min_gen);
+                        if (eq(*(powx->get_base()), *(powx2->get_base()))) {
+                            // won't handle things like 2**(3x/2) + 2**x, but
+                            // will handle 2**(x/2) + 2**x
+                            // TODO : can be modified to handle, will 
+                            // need some work
+                            // expand should not be needed here
+                            RCP<const Basic> tmp = expand(
+                                div(powx->get_exp(), powx2->get_exp()));
+                            int i = positive_integer(tmp);
+                            if (i > 0)
+                                continue;
+
+                            tmp = expand(
+                                div(powx2->get_exp(), powx->get_exp()));
+                            i = positive_integer(tmp);
+                            if (i > 0) {
+                                min_gen = curr_gen;
+                                continue;
+                            }
+                        }
+                    }
+
+                    throw std::runtime_error("Could not extract generator");
+                } else {
+                    throw std::runtime_error(
+                        "Internal Error : Wrong generator type");
+                }
+
+            } else {
+                SYMENGINE_ASSERT(is_a<const Symbol>(*curr_gen)
+                                 or is_a<const Pow>(*curr_gen)
+                                 or is_a_sub<const Function>(*curr_gen))
+                min_gen = curr_gen;
+                min_gen_set = true;
+            }
+        }
+
+        res = min_gen;
+    }
+
+    void bvisit(const Mul &x)
+    {
+        RCP<const Basic> gen;
+
+        if (not inpow)
+            if (not is_a<const Integer>(*x.coef_))
+                throw std::runtime_error("Non-integer coeff found");
+
+        if (x.dict_.size() > 1)
+            // this case should not be encountered, if `expand` was called
+            // apart from cases having more than one viable generator
+            throw std::runtime_error(
+                "Expand Error / Could not extract generator");
+
+        gen = _find_gen_uintpoly(
+            pow(x.dict_.begin()->first, x.dict_.begin()->second));
+
+        if (is_a<const Rational>(*x.coef_))
+            gen = mul(gen, div(one, rcp_static_cast<const Rational>(x.coef_)
+                                        ->get_den()));
+        res = gen;
+    }
+
+    void bvisit(const UIntPoly &x)
+    {
+        res = x.get_var();
+    }
+
+    void bvisit(const Function &x)
+    {
+        res = x.rcp_from_this();
+    }
+
+    void bvisit(const Symbol &x)
+    {
+        res = x.rcp_from_this();
+    }
+
+    void bvisit(const Basic &x)
+    {
+        throw std::runtime_error("Could not extract generator");
```

Comment:
There are other classes like `Constant`. It's simpler to have one method for `Basic`, which would handle `Function`, `Symbol`, `Constant`, etc.